src/aws/osml/model_runner/inference/http_detector.py

```diff
@@ -0,0 +1,109 @@
+from aws_embedded_metrics.unit import Unit
+from geojson import FeatureCollection
+
+from aws.osml.model_runner.api import ModelInvokeMode
+from aws.osml.model_runner.app_config import MetricLabels, ServiceConfig
+from aws.osml.model_runner.common import Timer
+
+from .detector import Detector
+from .endpoint_builder import FeatureEndpointBuilder
+from .feature_utils import create_mock_feature_collection
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPDetector(Detector):
+    def __init__(self, endpoint: str) -> None:
+        """
+        A HTTP model endpoint invoking object, intended to query sagemaker endpoints.
+
+        :param endpoint: str = the full URL to invoke the model
+
+        :return: None
+        """
+        self.http_pool = urllib3.PoolManager(cert_reqs="CERT_NONE")
+        super().__init__(endpoint=endpoint)
+
+    @property
+    def mode(self) -> ModelInvokeMode:
+        return ModelInvokeMode.HTTP_ENDPOINT
+
+    @metric_scope
+    def find_features(self, payload: BufferedReader, metrics: MetricsLogger) -> FeatureCollection:
+        """
+        Query the established endpoint mode to find features based on a payload
+
+        :param payload: BufferedReader = the BufferedReader object that holds the
+                                    data that will be sent to the feature generator
+        :param metrics: MetricsLogger = the metrics logger object to capture the log data on the system
+
+        :return: FeatureCollection = a feature collection containing the center point of a tile
+        """
+        retry_count = 0
+        logger.info("Invoking HTTP Endpoint: {}".format(self.endpoint))
+        if isinstance(metrics, MetricsLogger):
+            metrics.set_dimensions()
+            metrics.put_dimensions({"HTTPModelEndpoint": self.endpoint})
+
+        try:
+            self.request_count += 1
+            if isinstance(metrics, MetricsLogger):
+                metrics.put_metric(MetricLabels.MODEL_INVOCATION, 1, str(Unit.COUNT.value))
+
+            with Timer(
+                task_str="Invoke HTTP Endpoint",
+                metric_name=MetricLabels.ENDPOINT_LATENCY,
+                logger=logger,
+                metrics_logger=metrics,
+            ):
+                # If we are not running against a real model
+                if self.endpoint == ServiceConfig.noop_model_name:
+                    return create_mock_feature_collection(payload)
+                else:
+                    response = self.http_pool.request(
+                        method="POST",
+                        url=self.endpoint,
+                        body=payload,
+                    )
+                    self.request_count = 1
+                    if isinstance(metrics, MetricsLogger):
+                        metrics.put_metric(MetricLabels.ENDPOINT_RETRY_COUNT, retry_count, str(Unit.COUNT.value))
+                    return geojson.loads(response.data.decode("utf-8"))
```

Comment:
Shouldn't we be checking some HTTP status codes here? What happens if the server responds with some kind of 4xx / 5xx status code and a JSON formatted error message. It looks like we'll assume that was successful and continue on.

Reply:
This still needs to be addressed. The urllib3 ConnectionPool will pass through failed response codes that are not retryable (see: [lines 928-958 in connectionpool.py](https://github.com/urllib3/urllib3/blob/main/src/urllib3/connectionpool.py#L928)). We need to check for some kind of 2xx code before assuming the response was successful and update our error counts if it is not.